include/RAJA/pattern/forall.hpp

```diff
@@ -467,17 +504,22 @@ forall_Icount(ExecutionPolicy&& p,
  *
  ******************************************************************************
  */
-template <typename ExecutionPolicy, typename Res, typename Container, typename LoopBody>
+
+template <typename ExecutionPolicy, typename Res, typename Container, typename... Params>
 RAJA_INLINE concepts::enable_if_t<
     resources::EventProxy<Res>,
     concepts::negate<type_traits::is_indexset_policy<ExecutionPolicy>>,
     concepts::negate<type_traits::is_multi_policy<ExecutionPolicy>>,
     type_traits::is_range<Container>>
-forall(ExecutionPolicy&& p, Res r, Container&& c, LoopBody&& loop_body)
+forall(ExecutionPolicy&& p, Res r, Container&& c, Params... params)
 {
   static_assert(type_traits::is_random_access_range<Container>::value,
                 "Container does not model RandomAccessIterator");
 
+  auto f_params = expt::make_forall_param_pack(params...);
+  auto loop_body = expt::get_lambda(params...);
+  expt::check_forall_optional_args(loop_body, f_params);
```

Comment:
why is this the only version to do this check?

Reply:
good catch, calling this wherever we are doing this argument extraction

Reply:
What happens with forall_Icount, is it the different call signature on the call operator?